### PR TITLE
perf: Multiple nitpicks, some best practices and performances

### PR DESCRIPTION
**Pull request by FBruzzesi**
# Summary

So I might (or might not) have entered a rabbit hole, while the initial scope was much smaller (as you might tell from the branch name).

Some of the changes should lead to performance improvements, other are best practices.
Majority of them were spotted via [perflint](https://github.com/tonybaloney/perflint) or [refurb](https://github.com/dosisod/refurb).

There is no logic change, locally all tests pass without changing any of them.

---

If this is something of interest, feel free to help with the PR name, otherwise feel free to close it or cherry pick some of these changes.

# Related GitHub Issues and PRs

NA

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct/).
- [x] I have followed the [Style Guide for Python Code](https://peps.python.org/pep-0008/) as best as possible for the submitted code.
- [ ] I have added **pytest** unit tests for any new functionality.

### CODE REVIEW

File: great_tables/_utils_render_latex.py [hunk 1]
```diff
@@ -500,14 +489,14 @@ def derive_table_width_statement_l(data: GTData, use_longtable: bool) -> str:
         tw = float(tbl_width.strip("%"))
 
         side_width = (100 - tw) / 200
-        side_width = f"{side_width:.6f}".rstrip("0").rstrip(".")
+        side_width = f"{side_width:.6f}".rstrip(".0")
```
[the remark this answers is not in the document]
Comment (FBruzzesi) [reply great_tables/_utils_render_latex.py:492]: Good catch! You are right!
Repro: 

```py
"0.0".rstrip(".0")  # ''
"0.0".rstrip("0").rstrip(".")  # '0'
```

File: great_tables/_gt_data.py [hunk 2]
```diff
@@ -390,40 +386,23 @@ def align_from_data(self, data: TblData) -> Self:
...
         align: list[str] = []
 
+        align_to_left = {"object", "utf8", "string"}
         for col_class in col_classes:
             # Ensure that `col_class` is lowercase
             col_class = str(col_class).lower()
 
             # Translate the column classes to an alignment value of 'left', 'right', or 'center'
-            if col_class == "character-numeric":
+            if col_class == "character-numeric" or _str_detect(col_class, r"int|uint|float|date"):
```
Comment (machow) [inline great_tables/_gt_data.py:395]: @rich-iannone can you double check?

File: great_tables/_utils_nanoplots.py [hunk 3]
```diff
@@ -935,15 +932,8 @@ def _generate_nanoplot(
...
-    for i in range(len(y_proportions)):
-        y_proportions[i] = safe_y_d + ((1 - y_proportions[i]) * data_y_height)
-
-    for i in range(len(x_proportions)):
-        x_proportions[i] = (data_x_width * x_proportions[i]) + safe_x_d
-
-    data_y_points = y_proportions
-    data_x_points = x_proportions
+    data_y_points = tuple(safe_y_d + ((1 - y_p) * data_y_height) for y_p in y_proportions)
+    data_x_points = tuple((data_x_width * x_p) + safe_x_d for x_p in x_proportions)
```
Comment (machow) [inline great_tables/_utils_nanoplots.py:936]: It looks like x_proportions and y_proportions are mutated often in the original code. Can we restore this bit of code back?

Even if the tests pass, I want to make sure we do a bigger sweep first over the logic in this function, to be sure its defining / mutating / redefining things like `y_proportions` doesn't produce surprising edge cases (that might not be tested)

File: great_tables/_utils_render_latex.py [hunk 4]
```diff
@@ -295,49 +295,38 @@ def create_columns_component_l(data: GTData) -> str:
         #     ids=True,
         #     omit_columns_row=True,
         # )
-
-        for i in range(len(spanners)):
-            spanners_row = spanners[i]
-
-            for k, v in spanners_row.items():
-                if v is None:
-                    spanners_row[k] = ""
```
Comment (machow) [inline great_tables/_utils_render_latex.py:304]: Just noting this mutation that was fixed to not mutate (hopefully isn't leaned on elsewhere in code)

**Review by machow — APPROVED** [review]
This LGTM -- thanks for going over the code and making these changes. I do think it would be easier to review if this was split into a couple PRs, but at the same time the changes all seemed very reasonable, so I don't want to slow your roll 😁 

@rich-iannone do you mind looking over the align regex code I marked, just to double check it still matches how you think about the auto align code?

https://github.com/posit-dev/great-tables/pull/731/files#r2234015522

File: great_tables/_utils_render_latex.py (hunk 4 above)
Comment (FBruzzesi) [reply great_tables/_utils_render_latex.py:304]: Could you expand on this? I am not sure I understand what you mean

**Comment by FBruzzesi** [comment]
Thanks for the review @machow

> I do think it would be easier to review if this was split into a couple PRs, 

I agree, I even mentioned it in the description:

> So I might (or might not) have entered a rabbit hole, while the initial scope was much smaller (as you might tell from the branch name).

I can split it into different types of improvements

- performance
- best practices
- nitpicks

As also mentioned, it's ok if some changes are not liked 

> but at the same time the changes all seemed very reasonable, so I don't want to slow your roll 😁

I don't want to slow down or disencourange reviews by opening PRs that are too large, I will try to keep that in mind and have more contained changes

File: great_tables/_gt_data.py (hunk 2 above)
Comment (rich-iannone) [reply great_tables/_gt_data.py:395]: There's no difference in auto-alignment behavior from this PR compared to `main`. So everything looks alright to me.

File: great_tables/_utils_nanoplots.py (hunk 3 above)
Comment (machow) [reply great_tables/_utils_nanoplots.py:936]: Thanks for taking the time to look into how these are used. Let's just get this PR merged, since it seems like it shouldn't change things!

**Comment by machow** [comment]
Thanks for all your thoughtful work and response! These kinds of PRs are super helpful, so really appreciate you taking the time to improve so much of the code!
